### DISCUSSION TITLE
Build type definitions

### DISCUSSION THREAD

**Discussion by kentcdodds** [remix-run/remix#6270] [Proposals]
The `build/index.js` file generated by Remix's build doesn't have types associated. As a result, I'm getting a TypeScript error in my server when trying to follow the template by @jacob-ebey in https://github.com/remix-run/remix/pull/6229:

<img width="1010" alt="image" src="https://user-images.githubusercontent.com/1500684/235545230-aa19d9cf-e78d-43a9-9f21-3d711de8c231.png">

```
Type 'typeof import("/Users/kentcdodds/code/epicweb-dev/rocket-rental/build/index")' is not assignable to type 'ServerBuild'.
  The types of 'entry.module' are incompatible between these types.
    Property 'default' is missing in type '{}' but required in type 'ServerEntryModule'.ts(2322)
build.d.ts(32, 5): 'default' is declared here.
server.d.ts(18, 5): The expected type comes from property 'build' which is declared here on type '{ build: ServerBuild; getLoadContext?: GetLoadContextFunction | undefined; mode?: string | undefined; }'
```

Because the build is a `.js` file, TypeScript doesn't know what the types for it are. Would it be possible for Remix to write a `.d.ts` file next to the `build/index.js`?

**Comment by kentcdodds** [comment]
Alternatively, Remix could export a handy util to validate the `build` object is correct at runtime.

**Comment by pcattori** [comment]
Yes we already have plans for providing public types for main Remix artifacts (config, assets manifest, server build)
